File: apps/web/src/lib/api/activity.ts
import type { Route } from 'next';

import { apiFetch } from '@/lib/api/client';

export type ActivityAction =
  | 'created'
  | 'viewed'
  | 'updated'
  | 'status_changed'
  | 'archived'
  | 'restored'
  | 'deleted'
  | 'exported'
  | 'approved'
  | 'rejected'
  | 'login'
  | 'logout'
  | 'login_failed'
  | 'role_assigned'
  | 'role_removed'
  | 'permission_changed'
  | 'password_changed'
  | 'invited'
  | 'activated'
  | 'deactivated'
  | 'payment_completed'
  | 'funding_added'
  | 'file_uploaded'
  | 'note_added'
  | 'other';

export type ActivityEntityType =
  | 'user'
  | 'role'
  | 'lead'
  | 'investor'
  | 'project'
  | 'financial_account'
  | 'transaction'
  | 'project_budget'
  | 'funding_commitment'
  | 'payment_obligation'
  | 'document'
  | 'design_project';

export type ActivitySource =
  | 'web'
  | 'api'
  | 'background_job'
  | 'automation'
  | 'import'
  | 'integration'
  | 'ai_service';

export interface ActivityLogEntry {
  id: string;
  event_type: string;
  action: ActivityAction;
  entity_type: ActivityEntityType;
  entity_id: string;
  actor_type: string;
  actor_user_id: string | null;
  actor_name: string | null;
  source: ActivitySource;
  description_key: string;
  metadata: Record<string, unknown> | null;
  changed_fields: string[] | null;
  previous_values: Record<string, unknown> | null;
  new_values: Record<string, unknown> | null;
  ip_address: string | null;
  user_agent: string | null;
  request_id: string | null;
  is_demo: boolean;
  created_at: string;
  entity_label: string | null;
  link_module: string | null;
}

export interface ActivityListResponse {
  items: ActivityLogEntry[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export interface ActivityFilters {
  search: string;
  entity_type: ActivityEntityType | '';
  action: ActivityAction | '';
  source: ActivitySource | '';
  date_from: string;
  date_to: string;
  sort_order: 'asc' | 'desc';
  page: number;
  page_size: number;
}

export const EMPTY_ACTIVITY_FILTERS: ActivityFilters = {
  search: '',
  entity_type: '',
  action: '',
  source: '',
  date_from: '',
  date_to: '',
  sort_order: 'desc',
  page: 1,
  page_size: 25,
};

function buildQuery(filters: Partial<ActivityFilters>): string {
  const params = new URLSearchParams();
  if (filters.search) params.set('search', filters.search);
  if (filters.entity_type) params.set('entity_type', filters.entity_type);
  if (filters.action) params.set('action', filters.action);
  if (filters.source) params.set('source', filters.source);
  if (filters.date_from) params.set('date_from', `${filters.date_from}T00:00:00Z`);
  if (filters.date_to) params.set('date_to', `${filters.date_to}T23:59:59Z`);
  if (filters.sort_order) params.set('sort_order', filters.sort_order);
  if (filters.page) params.set('page', String(filters.page));
  if (filters.page_size) params.set('page_size', String(filters.page_size));
  const query = params.toString();
  return query ? `?${query}` : '';
}

export async function fetchActivities(filters: ActivityFilters = EMPTY_ACTIVITY_FILTERS) {
  return apiFetch<ActivityListResponse>(`/activity${buildQuery(filters)}`);
}

export async function fetchActivityDetail(activityId: string) {
  return apiFetch<ActivityLogEntry>(`/activity/${activityId}`);
}

export async function fetchEntityActivity(
  entityType: ActivityEntityType,
  entityId: string,
  limit = 20,
) {
  return apiFetch<ActivityListResponse>(
    `/activity/entity/${entityType}/${entityId}?limit=${limit}`,
  );
}

export function activityRecordHref(entry: ActivityLogEntry): Route | null {
  if (!entry.link_module) return null;
  return `/dashboard/${entry.link_module}` as Route;
}

export function formatActivityDate(value: string, locale: string): string {
  const intlLocale = locale === 'tr' ? 'tr-TR' : 'en-US';
  return new Intl.DateTimeFormat(intlLocale, {
    dateStyle: 'medium',
    timeStyle: 'short',
  }).format(new Date(value));
}

export function metadataForI18n(metadata: Record<string, unknown> | null | undefined) {
  if (!metadata) return {};
  return Object.fromEntries(
    Object.entries(metadata).map(([key, value]) => [key, value == null ? '' : String(value)]),
  ) as Record<string, string>;
}

File: apps/api/src/investhome_api/models/activity.py
import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, JSON, String, Uuid, func
from sqlalchemy.orm import Mapped, mapped_column

from investhome_api.db.base import Base


class ActivityActorType(str, enum.Enum):
    USER = "user"
    SYSTEM = "system"
    AI = "ai"
    INTEGRATION = "integration"


class ActivitySource(str, enum.Enum):
    WEB = "web"
    API = "api"
    BACKGROUND_JOB = "background_job"
    AUTOMATION = "automation"
    IMPORT = "import"
    INTEGRATION = "integration"
    AI_SERVICE = "ai_service"


class ActivityAction(str, enum.Enum):
    CREATED = "created"
    VIEWED = "viewed"
    UPDATED = "updated"
    STATUS_CHANGED = "status_changed"
    ARCHIVED = "archived"
    RESTORED = "restored"
    DELETED = "deleted"
    EXPORTED = "exported"
    APPROVED = "approved"
    REJECTED = "rejected"
    LOGIN = "login"
    LOGOUT = "logout"
    LOGIN_FAILED = "login_failed"
    ROLE_ASSIGNED = "role_assigned"
    ROLE_REMOVED = "role_removed"
    PERMISSION_CHANGED = "permission_changed"
    PASSWORD_CHANGED = "password_changed"
    INVITED = "invited"
    ACTIVATED = "activated"
    DEACTIVATED = "deactivated"
    PAYMENT_COMPLETED = "payment_completed"
    FUNDING_ADDED = "funding_added"
    FILE_UPLOADED = "file_uploaded"
    NOTE_ADDED = "note_added"
    OTHER = "other"


class ActivityEntityType(str, enum.Enum):
    USER = "user"
    ROLE = "role"
    LEAD = "lead"
    INVESTOR = "investor"
    PROJECT = "project"
    FINANCIAL_ACCOUNT = "financial_account"
    TRANSACTION = "transaction"
    PROJECT_BUDGET = "project_budget"
    FUNDING_COMMITMENT = "funding_commitment"
    PAYMENT_OBLIGATION = "payment_obligation"
    DOCUMENT = "document"
    COMPANY = "company"
    OFFICE = "office"
    DEPARTMENT = "department"
    TEAM = "team"
    BRAND = "brand"
    DESIGN_PROJECT = "design_project"
    BUILDING = "building"
    FLOOR = "floor"
    INVENTORY_ASSET = "inventory_asset"
    INVENTORY_RESERVATION = "inventory_reservation"
    INVENTORY_PRICE = "inventory_price"
    INVENTORY_OWNERSHIP = "inventory_ownership"
    INVENTORY_ASSIGNMENT = "inventory_assignment"
    PRICE_CHANGE_REQUEST = "price_change_request"
    SALES_OPPORTUNITY = "sales_opportunity"
    LEAD_QUALIFICATION = "lead_qualification"
    SALES_PROPOSAL = "sales_proposal"
    WORK_ITEM = "work_item"


class ActivityLog(Base):
    __tablename__ = "activity_logs"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    event_type: Mapped[str] = mapped_column(String(80), nullable=False)
    action: Mapped[ActivityAction] = mapped_column(
        Enum(ActivityAction, native_enum=False, length=50),
        nullable=False,
    )
    entity_type: Mapped[ActivityEntityType] = mapped_column(
        Enum(ActivityEntityType, native_enum=False, length=50),
        nullable=False,
    )
    entity_id: Mapped[uuid.UUID] = mapped_column(Uuid, nullable=False)
    actor_type: Mapped[ActivityActorType] = mapped_column(
        Enum(ActivityActorType, native_enum=False, length=50),
        nullable=False,
        default=ActivityActorType.SYSTEM,
    )
    actor_user_id: Mapped[uuid.UUID | None] = mapped_column(Uuid, nullable=True)
    actor_name: Mapped[str | None] = mapped_column(String(255), nullable=True)
    source: Mapped[ActivitySource] = mapped_column(
        Enum(ActivitySource, native_enum=False, length=50),
        nullable=False,
        default=ActivitySource.API,
    )
    description_key: Mapped[str] = mapped_column(String(120), nullable=False)
    metadata_json: Mapped[dict[str, object] | None] = mapped_column(JSON, nullable=True)
    changed_fields: Mapped[list[str] | None] = mapped_column(JSON, nullable=True)
    previous_values: Mapped[dict[str, object] | None] = mapped_column(JSON, nullable=True)
    new_values: Mapped[dict[str, object] | None] = mapped_column(JSON, nullable=True)
    ip_address: Mapped[str | None] = mapped_column(String(64), nullable=True)
    user_agent: Mapped[str | None] = mapped_column(String(500), nullable=True)
    request_id: Mapped[str | None] = mapped_column(String(64), nullable=True)
    is_demo: Mapped[bool] = mapped_column(default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
    )
